lab11.md

```diff
@@ -0,0 +1,184 @@
+HOME=/root
+Http_User_Agent=Go-http-client/1.1
+Http_Content_Length=0
+Http_Accept_Encoding=gzip
+...
+Http_Hmac=sha1=9239edfe20185eafd7a5513c303b03d207d22f64
+...
+```
+
+We see the `hmac` value transformed to environmental variable `Http_Hmac` with encrypted request `The message` with key `cookie` as `9239edfe20185eafd7a5513c303b03d207d22f64` and encryption method `sha1` appended in front of the value.
+
+## HMAC in action
+
+For our purpose we are going to create a new Python 3 function. Let’s call it `hmac-protected`:
+
+```
+$ faas-cli new --lang python3 hmac-protected --prefix="<your-docker-username>"
+```
+
+Add `payload-secret` which will serve as the key that will encrypt the payload. 
+
+Populate `payload-secret` like we did in [lab10](https://github.com/openfaas/workshop/blob/master/lab10.md):
+
+```
+$ echo -n "<your-secret>" | docker secret create payload-secret -
+```
+
+> Note: Remember the string you put in place of  `<your-secret>`
+
+Our `hmac-protected.yml` should look like:
+
+```
+provider:
+  name: faas
+  gateway: http://127.0.0.1:8080
+
+functions:
+  hmac-protected:
+    lang: python3
+    handler: ./hmac-protected
+    image: <your-docker-username>/hmac-protected:latest
+    secrets:
+      - payload-secret
+```
+
+Replace the content of the `handler.py` with the following code:
+
+```
+import os, hmac, hashlib
+
+def validateHMAC(message,secret,hash):
+
+    # The hash starts with the encryption method "sha1=" we need to remove it
+    cleanHash = hash[5:]
+    encodedSecret = secret.encode()
+    encodedMessage = message.encode()
+
+    # Encrypt message with secret
+    expectedMac = hmac.new(encodedSecret,encodedMessage,hashlib.sha1)
+    hmacDigest = expectedMac.hexdigest()
+
+    # Compare the received hash with created hash
+    if hmacDigest == cleanHash:
+        return True
+    return False
+
+def handle(req):
+    # We receive the encrypted message in form of a header
+    encryptedMessage = os.getenv("Http_Hmac")
+
+    # Read secret from inside the container
+    with open("/var/openfaas/secrets/payload-secret","r") as secretContent:
+        payloadSecret = secretContent.read()
+
+    # Function to validate the HMAC
+    if validateHMAC(req,payloadSecret,encryptedMessage):
+        return "Successfully validated: " + req
+    return "HMAC validation failed."
+```
+
+> The source code is also available at [hmac-protected/hmac-protected/handler.py](./hmac-protected/hmac-protected/handler.py)
+
+* Build, push and deploy the function:
+
+```
+$ faas up -f ./hmac-protected.yml
+```
+
+### Invoke function
+
+We will invoke the function by sending two values:
+
+* The normal request message
+
+* A header sent with `--sign` containing the same message, but encrypted with the value of the `--key` label
+
+then inside the function we will receive the values from above. The normal message will be encrypted with the value of `payload-secret` which will be the same value as the `--key` and the hashes compared:
```

Comment:
Better use capital letter

Reply:
You can edit the last sentence, as this doesn't sound clear enough ( literally says that `payload-secret` = comparison between `--key` and the hash.